### DISCUSSION TITLE
Pixelbook Go i7 KVM support

### DISCUSSION THREAD

**Discussion by hopgoodpaul** [sebanc/brunch#792] [Q&A]
I'm running a Pixelbook Go i7 model with Mr Chromebox's UEFI firmware. I'm trying to run an updated kernel so that I can take advantage of the nested KVM support.

I've used Brunch to dual boot linux mint and Chrome OS atlas image which works nicely for the most part. Audio, microphone and camera don't appear to work even after trying a couple of different kernel options.

Am I right in thinking that crunch might be a better option for me? From reading the readme, I can see it says it uses the native device kernel for Chrome OS with some modifications. My question is: do those modifications mean nested KVM is available?

I need to have audio, mic and camera working as well as KVM support. 

Thanks for any help

**Comment by sebanc** [comment]
Crunch uses the official kernel for your device so it will not enable nested KVM.

Brunch should however do that as it uses a more recent kernel but the specific firmwares for your device would need to be identified and added in brunch as a new option (same as what "asus_c302" option does).

Most of the firmware are located in the recovery image partition 3 (which can only be mounted read only) at the following places:
- /lib/firmware
- /usr/share/alsa/ucm
- /etc/camera

So that I can add support for Pixelbook Go, you need to identify the the missing firmware files, copy them to the brunch rootfs and check if sound / camera start working.

If you list the firmware files which are needed for Pixelbook Go, I can probably add support for it.

**Comment by hopgoodpaul** [comment]
I've had a few different tries at this over the weekend and haven't had any success unfortunately. 

I believe I've identified all the firmware files and .conf files from those directories. It's hard to say for sure. 
I've copied them from the original partition 3 that was still present after flashing the UEFI rom and booting from USB to Mint.

I've updated the packages.sh script to add an option for pixelbook_go by copying the lines doing the same for asus_c302. I also included the firmware files in a tar.gz file to match the asus_c302 option.

I've tried using different kernel versions to see if that made any difference as well.

dmesg still keeps showing "ALSA device list: no soundcards found".

It's possible I'm still missing some firmware files or I've done something wrong when copying the files from the original partition.

I'll keep trying a few different things around those firmware files.

**Comment by sebanc** [comment]
You seem to be on the right track :)

I forgot to mention that you would need to add the option "disable_intel_hda" or to blacklist snd_hda_intel module some other way for the sound driver to work.

Could you post a dmesg with that option activated or with snd_hda_intel module blacklisted some other way ? (run "dmesg > ~/Downloads/dmesg.txt" in crosh shell and attach dmesg.txt from your downloads folder)

Note: The "ALSA device list: no soundcards found" message is normal on early boot, it just means that sound drivers are built as modules and not builtin.

**Reply by hopgoodpaul** [reply]
Thanks sebanc

I read on a few pages adding "disable_intel_hda" so have already got that in as an option.

I've created the dmesg.txt file as requested and attached it here. I did notice the section below while searching through the file. Right now I'm not sure what would be stopping the firmware files from being loaded in. Could this be the problem?

```
[   96.637030] snd_soc_skl 0000:00:1f.3: Direct firmware load for 9d71-GOOGLE-ATLASMAX-0-tplg.bin failed with error -2
[   96.637043] snd_soc_skl 0000:00:1f.3: tplg fw 9d71-GOOGLE-ATLASMAX-0-tplg.bin load failed with -2, trying alternative tplg name kbl_da7219_max98373-tplg.bin
[   96.637462] snd_soc_skl 0000:00:1f.3: Direct firmware load for kbl_da7219_max98373-tplg.bin failed with error -2
[   96.637470] snd_soc_skl 0000:00:1f.3: tplg kbl_da7219_max98373-tplg.bin failed with -2, falling back to dfw_sst.bin
[   96.637491] snd_soc_skl 0000:00:1f.3: Direct firmware load for dfw_sst.bin failed with error -2
[   96.637497] snd_soc_skl 0000:00:1f.3: Fallback tplg fw dfw_sst.bin load failed with -2
[   96.637502] snd_soc_skl 0000:00:1f.3: Failed to init topology!
[   96.637507] snd_soc_skl 0000:00:1f.3: ASoC: error at snd_soc_component_probe on 0000:00:1f.3: -2
[   96.637512] snd_soc_skl 0000:00:1f.3: ASoC: failed to probe component -2
[   96.637543] kbl_da7219_max98_927_373 kbl_da7219_max98373: ASoC: failed to instantiate card -2
[   96.637835] kbl_da7219_max98_927_373: probe of kbl_da7219_max98373 failed with error -2
```

[dmesg.txt](https://github.com/sebanc/brunch/files/5798495/dmesg.txt)

Thanks for the help. It feels a step closer to working 👍

**Reply by sebanc** [reply]
Indeed try to see if there is a fw named 9d71-GOOGLE-ATLASMAX-0-tplg.bin in /lib/firmware in the atlas recovery image and copy it to brunch rootfs

**Comment by hopgoodpaul** [comment]
Ok I've added that missing bin plus a couple of others. I also fixed some permissions errors I was getting as well.

Attached is the latest dmesg output.

[dmesg120121.txt](https://github.com/sebanc/brunch/files/5800294/dmesg120121.txt)

When searching for "snd" there are a few entries that now come up that I don't really understand:

```[   81.433311] pci 0000:00:1f.3: attach allowed to drvr snd_soc_skl [trusted device]
[   81.433341] snd_soc_skl 0000:00:1f.3: DSP detected with PCI class/subclass/prog-if info 0x040100
[   82.540660] snd_soc_skl 0000:00:1f.3: bound 0000:00:02.0 (ops intel_dual_link_lvds_callback [i915])
[   82.567568] snd_soc_skl 0000:00:1f.3: ASoC: old version of manifest
[   82.568228] snd_soc_skl 0000:00:1f.3: ASoC: Parent card not yet available, widget card binding deferred
[   82.617740] snd_soc_skl 0000:00:1f.3: skl_tplg_add_moduleid_in_bind_params: invalid kpb post bind params
[   82.660492] kbl_da7219_max98_927_373 kbl_da7219_max98373: snd_soc_bind_card: snd_soc_dapm_add_routes failed: -19
```

**Reply by hopgoodpaul** [reply]
Interestingly, I can get audio working when I plug in an external mic and speaker but when I plug in an external webcam, it doesn't work (I've confirmed the webcam functions properly on another machine and the mic works when plugged into the pixelbook).

It seems like I'm probably still missing some firmware files for the webcam.

I'm not entirely sure what to do about the internal sound either.

**Reply by sebanc** [reply]
What's the output of "sudo alsaucm reload" in crosh shell ?

**Reply by hopgoodpaul** [reply]
```
ALSA lib /build/atlas/tmp/portage/media-libs/alsa-lib-1.2.1.2-r2/work/alsa-lib-1.2.1.2/src/ucm/parser.c:1741:(uc_mgr_scan_master_configs) error: could not scan directory /usr/share/alsa/ucm2: No such file or directory
alsaucm: unable to obtain card list: No such file or directory
```

**Reply by hopgoodpaul** [reply]
A quick search online landed me here: https://bugs.archlinux.org/task/67027
I saved the conf file and created the ucm2 folder.

Now the output from `sudo alsaucm reload` is simply "No card found"

**Reply by sebanc** [reply]
What's the output of "alsactl init" ?

**Reply by hopgoodpaul** [reply]
```
alsactl: sysfs_init:48: sysfs path '/sys' is invalid

alsactl: init:1759: No soundcards found...
```

**Reply by sebanc** [reply]
Also, could you verify if you have copied those files (which are also present in the atlas recovery image) to /usr/share/alsa/ucm folder:
https://chromium.googlesource.com/chromiumos/overlays/board-overlays/+/refs/heads/main/overlay-atlas/chromeos-base/chromeos-bsp-atlas/files/audio-config/ucm-config/kblda7219m98373

**Reply by sebanc** [reply]
Sorry and the output of "sudo alsactl init"

**Reply by hopgoodpaul** [reply]
Thanks for your continued help.
I can confirm those files are in the ucm folder as described.

The output of alsactl init is as below:

```
chronos@localhost / $ sudo alsactl init
alsactl: sysfs_init:48: sysfs path '/sys' is invalid

alsactl: init:1759: No soundcards found...
```

**Reply by sebanc** [reply]
Could you remove /usr/share/alsa/ucm2 folder and post a new dmesg ?

**Reply by hopgoodpaul** [reply]
Done and attached :)

[dmesg3.txt](https://github.com/sebanc/brunch/files/5804528/dmesg3.txt)

**Comment by sebanc** [comment]
Also, could you copy /lib/firmware/intel/dsp_fw_release.bin from the recovery image to the brunch rootfs ?

**Reply by hopgoodpaul** [reply]
Hmm that particular file doesn't appear to be in the recovery image I flashed to a usb drive. I'll keep looking for a copy of it.

**Reply by hopgoodpaul** [reply]
These are the files I do have in place.
```
dsp_fw_578A29B8-0890-446F-A80A-2FE202AE0DBA.bin  
dsp_fw_kbl.bin
dsp_fw_B489C2DE-0F96-42E1-8A2D-C25B5091EE49.bin  
dsp_fw_kbl_v1037.bin
dsp_fw_C75061F3-F2B2-4DCC-8F9F-82ABB4131E66.bin  
dsp_fw_kbl_v2042.bin
dsp_fw_bxtn_v2219.bin                            
dsp_fw_kbl_v2630.bin
dsp_fw_bxtn_v3366.bin                            
dsp_fw_kbl_v3266.bin
dsp_fw_cnl_v1191.bin                             
dsp_fw_kbl_v3402.bin
dsp_fw_cnl_v1858.bin                             
dsp_fw_kbl_v3420.bin
dsp_fw_glk_v1814.bin                             
dsp_fw_kbl_v701.bin
dsp_fw_glk_v2768.bin                             
dsp_fw_release_v3402.bin
dsp_fw_glk_v2880.bin                             
dsp_fw_release_v969.bin
dsp_fw_glk_v3366.bin
```

and these are the files I can have in the flash drive recovery image:
```
dsp_fw_578A29B8-0890-446F-A80A-2FE202AE0DBA.bin
dsp_fw_B489C2DE-0F96-42E1-8A2D-C25B5091EE49.bin
dsp_fw_C75061F3-F2B2-4DCC-8F9F-82ABB4131E66.bin
dsp_fw_kbl.bin
dsp_fw_kbl_v1037.bin
dsp_fw_kbl_v2630.bin
dsp_fw_kbl_v3266.bin
dsp_fw_kbl_v3420.bin
```

**Reply by hopgoodpaul** [reply]
https://chromium.googlesource.com/chromiumos/third_party/linux-firmware/+/refs/heads/stabilize-10569.B/intel/

By the looks of this link, I should have a sym link from dsp_fw_release.bin pointing to dsp_fw_release_v969.bin

Should I create one?

**Reply by sebanc** [reply]
Yes try creating one

**Reply by sebanc** [reply]
Could you also verify that the below files from /lib/firmware have also been copied ?

dsp_lib_dsm_core_spt_release.bin
dsp_lib_ghw_spt_release.bin
dsp_lib_waves_spt_release.bin

**Reply by hopgoodpaul** [reply]
Ok I've created the sym link for dsp_fw_release pointing to dsp_fw_release_v969.bin. No change for audio etc.

I've created another sym link pointing dsp_fw_kbl.bin to dsp_fw_kbl_v3420.bin based on this link https://packages.debian.org/sid/firmware-intel-sound. Kinda of clutching at straws with it though.

I've verified that those files:
dsp_lib_dsm_core_spt_release.bin
dsp_lib_ghw_spt_release.bin
dsp_lib_waves_spt_release.bin
are all present on the ROOT-A partition (partition 3) of the Chrome OS install.

Here's the latest dmesg output file too
[dmesg4.txt](https://github.com/sebanc/brunch/files/5808215/dmesg4.txt)

I still get that same `'/sys' is invalid alsactl: init:1759: No soundcards found`  error message on `sudo alsactl init` too.

I'm thinking about starting fresh and copying nearly everything over from /etc, /lib and /usr from the working, native install of ChromeOS. Feels a bit extreme but might allow me to try things a little more tidily.

**Reply by hopgoodpaul** [reply]
I started again and copied all the files we've talked about making sure to preserve all the symlinks and files those links point to as well.
There's something that's stopping the audio and camera from working and really I'm beyond the limit of my knowledge of this stuff at the moment.

I was wondering about copying over the kernel files from /lib/modules etc. but, like I said, it's not really something I understand.

I really appreciate the help you've given.

Attached is the latest dmesg file in case you're interested. There not a huge amount of information I can find on the errors like 
```
[   81.847390] kbl_da7219_max98_927_373 kbl_da7219_max98373: snd_soc_bind_card: snd_soc_dapm_add_routes failed: -19
```
and I don't really know if they are related to my problem at all either.

Anyway, thanks again 👍 

[dmesg5.txt](https://github.com/sebanc/brunch/files/5819896/dmesg5.txt)

**Reply by sebanc** [reply]
I used the exact same steps to make audio work on my asus C302 so I am not sure what's missing here...

You could try another kernel version by replacing /kernel in the grub config with /kernel-4.19 or /kernel-5.10 (note that it will rebuild rootfs).

Also for camera, renaming 99-camera.sh to 99-camera.old in /patches directory in partition 7 should probably work (provided that you use the atlas recovery image).

**Comment by sebanc** [comment]
I added the option "native_chromebook_image" in the latest brunch-unstable build. This option is for debugging purpose mainly, it will leave all the original firmwares in the recovery image in place. Could you try to boot the new brunch-unstable build with "options=native_chromebook_image" and see if sound works ?

**Comment by hopgoodpaul** [comment]
Thanks for this. I've given it a try from the unstable build plus a few different options too. Each different boot gave the same results of no audio, mic or camera.

I checked dmesg after each boot and saved the output to different files. Attached here.
[dmesg-native_chromebook_image+disable_intel_hda+kernel510.txt](https://github.com/sebanc/brunch/files/5826858/dmesg-native_chromebook_image%2Bdisable_intel_hda%2Bkernel510.txt)
[dmesg-native_chromebook_image+disable_intel_hda+kernel54.txt](https://github.com/sebanc/brunch/files/5826859/dmesg-native_chromebook_image%2Bdisable_intel_hda%2Bkernel54.txt)
[dmesg-native_chromebook_image+disable_intel_hda.txt](https://github.com/sebanc/brunch/files/5826860/dmesg-native_chromebook_image%2Bdisable_intel_hda.txt)
[dmesg-native_chromebook_image.txt](https://github.com/sebanc/brunch/files/5826861/dmesg-native_chromebook_image.txt)

I noticed on this build you had already disabled/renamed the 99-camera.sh file so I didn't change that. 

My guess is there's something specific in the kernel that's missing.

**Reply by hopgoodpaul** [reply]
As I have a look around more, it's looking more likely that kernels later than 4.4 for kabylake have the same audio issues. It's making me thing that perhaps one route would be to try [disabling module_locking](https://github.com/divx118/crouton-packages/blob/master/README.md) and then [enable nested kvm](https://www.kernel.org/doc/html/latest/virt/kvm/running-nested-guests.html) that way.

**Comment by sebanc** [comment]
I found 2 kernel configs which might be related to camera and audio issues, I will rebuild the kernel tonight and post a new test build.

**Comment by sebanc** [comment]
Could you try the latest brunch-unstable ?

**Comment by hopgoodpaul** [comment]
Ok, I've done a bit of testing with no luck so far but with different
options in the grub config, I am seeing different messages in dmesg.

I'll post my findings in the morning when I've done a bit more testing.

Thanks

**Comment by hopgoodpaul** [comment]
I've been doing a load of testing of a few different grub options with no change to audio and camera as yet but the dmesg output has changed on each different option.

I also get a different message from `alsactl init` too.

```
alsactl init
alsactl: sysfs_init:48: sysfs path '/sys' is invalid

Found hardware: "HDA-Intel" "Intel Kabylake HDMI" "HDA:8086280b,80860101,00100000" "0x1ae0" "0x006c"
Hardware is initialized using a generic method
```
So far, with my limited knowledge, it seems like the default grub file (with no options added at all) produces the least entries when searching for "snd" and no apparent error messages.

[dmesg.txt](https://github.com/sebanc/brunch/files/5837002/dmesg.txt)

[dmesg+kernel510+chromebook_native_image+disable_intel_hda.txt](https://github.com/sebanc/brunch/files/5836992/dmesg%2Bkernel510%2Bchromebook_native_image%2Bdisable_intel_hda.txt)
[dmesg+kernel510+chromebook_native_image.txt](https://github.com/sebanc/brunch/files/5836993/dmesg%2Bkernel510%2Bchromebook_native_image.txt)
[dmesg+kernel510.txt](https://github.com/sebanc/brunch/files/5836994/dmesg%2Bkernel510.txt)
[dmesg+kernel54+native_chromebook_image+disable_intel_hda.txt](https://github.com/sebanc/brunch/files/5836995/dmesg%2Bkernel54%2Bnative_chromebook_image%2Bdisable_intel_hda.txt)
[dmesg+kernel54+native_chromebook_image.txt](https://github.com/sebanc/brunch/files/5836996/dmesg%2Bkernel54%2Bnative_chromebook_image.txt)
[dmesg+kernel54+disable_intel_hda.txt](https://github.com/sebanc/brunch/files/5836997/dmesg%2Bkernel54%2Bdisable_intel_hda.txt)
[dmesg+kernel54.txt](https://github.com/sebanc/brunch/files/5836998/dmesg%2Bkernel54.txt)
[dmesg+disable_intel_hda+native_chromebook_image.txt](https://github.com/sebanc/brunch/files/5836999/dmesg%2Bdisable_intel_hda%2Bnative_chromebook_image.txt)
[dmesg+disable_intel_hda.txt](https://github.com/sebanc/brunch/files/5837000/dmesg%2Bdisable_intel_hda.txt)
[dmesg+native_chromebook_image.txt](https://github.com/sebanc/brunch/files/5837001/dmesg%2Bnative_chromebook_image.txt)

**Reply by sebanc** [reply]
Ok so the change I applied apparently had no effect.

When you boot with the default grub file there is no error because the kernel is not even attempting to load the sound module (you need disable_intel_hda or native_chromebook_image for it to load).

Most of my assumptions regarding sound on chromebooks are based on this https://github.com/flantel/pixelbook-linux. Apparently eve can have sound in linux using the same chromiumos kernel release that I use in brunch:
https://github.com/flantel/pixelbook-linux/blob/master/ansible/roles/eve-kernel/defaults/main.yml

The main differences that I see are that this method uses the same 4.4 kernel as the original "eve" device and also the original kernel config.

If you could provide me with the atlas original kernel config, I could try a kernel build with it.

**Comment by hopgoodpaul** [comment]
Sure thing. What's the best way to get the kernel config out in the first
place?

**Reply by hopgoodpaul** [reply]
I've extracted the entire root-a partition to a .img file if that's any help?

**Reply by sebanc** [reply]
Actually I wish there would be a simple way to get it but the only way I know is to run the below command under the official chromeos in dev mode:
`sudo modprobe configs; zcat /proc/config.gz > ~/Downloads/atlas.config`

Actually if you can go back temporarily to the official chromeos on your device, there are 2 things that would be interesting:
1) get the kernel config
2) Enable legacy usb boot and try to boot a brunch usb flashdrive created with the mbr support patch (and with "native_chromebook_image" option) and get a dmesg to see if the issue  is the same

**Reply by hopgoodpaul** [reply]
I've got the kernel config as requested but I couldn't boot into chromeos on the USB flashdrive. 

Renamed to .txt because github doesn't support .config files.
[atlas-brunch.config.txt](https://github.com/sebanc/brunch/files/5849898/atlas-brunch.config.txt)
[atlas.config.txt](https://github.com/sebanc/brunch/files/5849900/atlas.config.txt)

I created a usb boot drive using the mbr support patch and the native_chromebook_image option but couldn't get it to boot.

If I use mrchromebox's firmware utility to install RW_LEGACY, I can get to the grub bootloader screen on the USB flashdrive, but when I try to boot into Chrome OS, I get an error telling me to load the kernel first. I can boot to linux mint on a flashdrive from RW_LEGACY though.

If I flash the firmware with UEFI, again, I can boot to linux mint on the flashdrive, but I get an error when trying to boot to the chromeos flashdrive. So I haven't been able to get the dmesg output for you.

**Reply by sebanc** [reply]
Thanks a lot for the kernel config, I will try to build a 4.4 kernel with this config to see if at least sound works with the original kernel version / config.

If it does, we can try to move forward with more recent kernel versions.

**Comment by hopgoodpaul** [comment]
No problem at all. Thank you again for all your help on this.

**Reply by sebanc** [reply]
Could you try the new release in brunch-unstable repository with the below additional steps:
- Replace "/kernel" in the grub configuration with "/kernel-test"
- add "options=native_chromebook_image" to the kernel command line

**Reply by hopgoodpaul** [reply]
No luck with those options unfortunately. Trackpad and touch screen weren't working and it wouldn't let me sign in. I enter my email, password and 2fa but then it sits just spinning and saying "Please wait".

I was able to press ctrl+alt+F2 to get to the shell and extract a dmesg log. 
[dmesg.txt](https://github.com/sebanc/brunch/files/5859002/dmesg.txt)

I'll try a couple of boot options tomorrow to see if anything helps with that.

**Reply by sebanc** [reply]
Thanks for the dmesg :) 

Don't bother trying more things with this build, I must have forgotten to add some swtpm configs and anyway the sound related logs are exactly the same.

I will try something else

**Reply by sebanc** [reply]
Could you try the new release in brunch-unstable repository with the same additional steps:
- Replace "/kernel" in the grub configuration with "/kernel-test"
- add "options=native_chromebook_image" to the kernel command line

**Reply by hopgoodpaul** [reply]
Never has the Chrome OS initial setup tune sounded so good. 

Audio and mic works but camera doesn't.

Excellent progress! 😃 

Dmesg attached
[dmesg.txt](https://github.com/sebanc/brunch/files/5860032/dmesg.txt)

**Reply by sebanc** [reply]
Nice to see some improvements :) This version has the 4.4 kernel with minimum patches so either the issue is that audio support has not been ported to more recent kernels or that one of my patches is responsible for that.

However, before attempting to get audio working on a more recent kernel, I would like to try fixing the camera. Unfortunately, I don't know much about ipu3 cameras, I have verified that the kernel contains the necessary configs / drivers but I see no log about a camera being loaded (it might be normal).

Maybe we could try things in this order:
1) Mount partition 7 and replace the content of /patches/99-disable_fw_updates.sh with the below one:
```
# Disable chromebooks firmware writing mechanisms to avoid the risk (very low) that ChromeOS would flash the firmware of an unsupported device

ret=0
echo "exit 0" > /roota/usr/sbin/chromeos-firmwareupdate
if [ ! "$?" -eq 0 ]; then ret=$((ret + (2 ** 1))); fi
echo "exit 0" > /roota/usr/sbin/flashrom
if [ ! "$?" -eq 0 ]; then ret=$((ret + (2 ** 3))); fi
exit $ret
```
Then reboot (it will rebuild the rootfs) and test the camera.

2) If it does not help, I would need a "dmesg" from the original chromeos with the original fw to see if it is supposed to show some logs about the camera or not.

**Reply by sebanc** [reply]
Actually, I have just noticed a major difference between atlas factory kernel and the generic 4.4, I will try something.

**Comment by sebanc** [comment]
I uploaded a new release in brunch-unstable (I have just updated the attached file, the release name is still the same actually). Could you try it ?

Forgot to mention that I changed the name of the kernel filename to "/kernel-chromebook" (instead of /kernel-test)

**Reply by hopgoodpaul** [reply]
Trying it now. You said you updated the attached file. Should something be attached here or is it in the brunch-unstable release? I only ask because the release time on github still says 6 hours ago.

**Reply by sebanc** [reply]
Exactly, the release still says 6 hours old but I changed the attached file

**Reply by hopgoodpaul** [reply]
I thought that's what you meant. I'm not seeing any file attached here though. Am I missing something?

**Reply by sebanc** [reply]
I meant I changed the file directly in brunch-unstable release: https://github.com/sebanc/brunch-unstable/releases/download/r87-unstable-20210123/brunch_r87_unstable_20210123_v2.tar.gz

**Reply by hopgoodpaul** [reply]
Ah ok got it! Thanks :)  Trying now

**Comment by hopgoodpaul** [comment]
[dmesg.txt](https://github.com/sebanc/brunch/files/5860520/dmesg.txt)

Ok, latest dmesg attached. Audio and mic still working but still no camera.

Should I try what you said earlier?

> Mount partition 7 and replace the content of /patches/99-disable_fw_updates.sh with the below one:
> #Disable chromebooks firmware writing mechanisms to avoid the risk (very low) that ChromeOS would flash the firmware of an unsupported device
> 
> ret=0
> echo "exit 0" > /roota/usr/sbin/chromeos-firmwareupdate
> if [ ! "$?" -eq 0 ]; then ret=$((ret + (2 ** 1))); fi
> echo "exit 0" > /roota/usr/sbin/flashrom
> if [ ! "$?" -eq 0 ]; then ret=$((ret + (2 ** 3))); fi
> exit $ret
> Then reboot (it will rebuild the rootfs) and test the camera.

**Reply by sebanc** [reply]
Could you run "sudo ln -s /lib/firmware/intel/ipu3-fw.bin /lib/firmware/ipu3-fw.bin", reboot and post a new dmesg ?

**Reply by hopgoodpaul** [reply]
Attached now.
[dmesg2.txt](https://github.com/sebanc/brunch/files/5860561/dmesg2.txt)


Interestingly as I test the mic a bit more using the Google Meet android app, it appears to work for a short while, then hangs and the app crashes. I think you can that in the dmesg file attached as well.

**Reply by hopgoodpaul** [reply]
It also looks like bluetooth isn't working either. It's on but doesn't show devices I can pair with.

**Reply by sebanc** [reply]
Ok the change I made did not help much, for both issues could you try the below and post an official chromeos dmesg if it does not help ?
```
Mount partition 7 and replace the content of /patches/99-disable_fw_updates.sh with the below one:
#Disable chromebooks firmware writing mechanisms to avoid the risk (very low) that ChromeOS would flash the firmware of an unsupported device

ret=0
echo "exit 0" > /roota/usr/sbin/chromeos-firmwareupdate
if [ ! "$?" -eq 0 ]; then ret=$((ret + (2 ** 1))); fi
echo "exit 0" > /roota/usr/sbin/flashrom
if [ ! "$?" -eq 0 ]; then ret=$((ret + (2 ** 3))); fi
exit $ret
Then reboot (it will rebuild the rootfs) and test the camera.
```

**Reply by hopgoodpaul** [reply]
Done that and attached the dmesg. Still no camera unfortunately.

[dmesg3.txt](https://github.com/sebanc/brunch/files/5860635/dmesg3.txt)

I'll restore to full native chrome os and post the dmesg from that. It'll take a bit of time though.

**Comment by hopgoodpaul** [comment]
Fully restored and back to native Chrome OS image now.

dmesg file attached
[dmesg.txt](https://github.com/sebanc/brunch/files/5860944/dmesg.txt)

**Reply by sebanc** [reply]
Thanks for the dmesg :) Could you try the new brunch-unstable release when you have the time ?

**Reply by hopgoodpaul** [reply]
Back on the latest brunch-unstable release and attached the latest dmesg.
[dmesg.txt](https://github.com/sebanc/brunch/files/5862781/dmesg.txt)

Still no camera working unfortunately. I try to have a poke around the dmesg output but I don't see anything obvious in it.

Is it worth me trying a few different options in the grub config?

**Reply by sebanc** [reply]
Could you try to apply the patch I previously posted for /patches/99-disable_fw_updates.sh and to delete /patches/99-preui_probe.sh (on partition 7) ?

**Comment by hopgoodpaul** [comment]
[dmesg2.txt](https://github.com/sebanc/brunch/files/5862992/dmesg2.txt)
Latest dmesg here.
No change, I'm afraid.
Thanks for keeping on looking into this!

**Reply by sebanc** [reply]
This modification had the effect I expected, rearranging modules loading order but unfortunately it does not seem to be related.

Could you post the /var/log/messages file ?

**Comment by hopgoodpaul** [comment]
Here you go :)
[messages.txt](https://github.com/sebanc/brunch/files/5863033/messages.txt)

**Reply by hopgoodpaul** [reply]
In case it's easier, I've joined the discord server and am online now.

**Comment by sebanc** [comment]
I have just updated the latest brunch-unstable archive (not the release) if you want to try it:
https://github.com/sebanc/brunch-unstable/releases/download/r87-unstable-20210124/brunch_r87_unstable_20210124_v2.tar.gz

**Reply by hopgoodpaul** [reply]
No luck unfortunately. It just hangs after accepting the 2fa with no working trackpad as before.

Here's the dmesg extract though.
[dmesg.txt](https://github.com/sebanc/brunch/files/5865148/dmesg.txt)

**Reply by hopgoodpaul** [reply]
And here is the messages file just in case you need it.
[messages.txt](https://github.com/sebanc/brunch/files/5865183/messages.txt)

**Reply by sebanc** [reply]
Could you try the new brunch-unstable ?

**Reply by hopgoodpaul** [reply]
Got logged in and the trackpad is working but no sound, mic or camera.

Here's the latest dmesg and messages log files.
[messages.txt](https://github.com/sebanc/brunch/files/5871995/messages.txt)
[dmesg.txt](https://github.com/sebanc/brunch/files/5871996/dmesg.txt)

**Reply by sebanc** [reply]
Could you try this build when you have the time ?
https://github.com/sebanc/brunch-unstable/releases/download/r87-unstable-20210126/brunch_r87_unstable_20210126_v2.tar.gz

**Reply by hopgoodpaul** [reply]
No change on the audio, mic and camera that I can see unfortunately.

[messages.txt](https://github.com/sebanc/brunch/files/5876422/messages.txt)
[dmesg.txt](https://github.com/sebanc/brunch/files/5876423/dmesg.txt)

I'll try deleting the 99-preui_probe.sh file now.

**Reply by hopgoodpaul** [reply]
No change but here are the log files if it's any help.
Anything else I should try changing?

[messages.txt](https://github.com/sebanc/brunch/files/5876466/messages.txt)
[dmesg.txt](https://github.com/sebanc/brunch/files/5876467/dmesg.txt)

**Reply by sebanc** [reply]
It does not seem to be a kernel config issue so I tried a patch in the new unstable build, could you try it ?

**Reply by hopgoodpaul** [reply]
Sure thing. I've given it a try but I'm afraid audio and camera still aren't working.

Here are the dmesg and messages files .
[messages.txt](https://github.com/sebanc/brunch/files/5882861/messages.txt)
[dmesg.txt](https://github.com/sebanc/brunch/files/5882862/dmesg.txt)

**Reply by hopgoodpaul** [reply]
Also, bluetooth isn't working either. I keep forgetting to check that one.

**Reply by sebanc** [reply]
I think I understood the sound issue with recent kernels. The sound driver "kbl_da7219_max98373" (in /sound/soc/intel/boards in the kernel tree) has been replaced by a SOF driver for which we don't have the correct firmwares. There have been significant changes in the sound kernel tree so porting the old driver would be very complicated for me. Unless someone ports this driver to more recent kernel versions, the only option I can see is to use the 4.4 kernel branch.

Regarding the camera, I have tried all I could think of, it can be related to:
- The chromiumos kernel source not being correct.
- an issue with the custom firmware.
- something else I am missing but I don't see what I could change to make it work.

I am going to build a final version of the 4.4 kernel with audio working. I did not really look into the bluetooth issue yet, if you still have the issue with 4.4 kernel, I can try to have a look at it.

**Reply by hopgoodpaul** [reply]
I understand and really appreciate the work you've put into this for me and I would be more than happy to donate something to help support what you've been doing.

I will try kernel 4.4 and see how well that works. 

For my particular use case (the main thing I want is to have nested kvm working), it seems like I should be able to use the genuine native atlas kernel and then modify that to enable nested kvm with a modprobe command. Is it a case of disabling rootfs verification and running the command or is there something else I'm missing?

Thanks again for your time.

**Reply by sebanc** [reply]
I have just uploaded the latest brunch-unstable, it has a 4.4 chromebook kernel and I expect it to work fine aside from cameras. Let me know how it goes with bluetooth.
Also, could you try to update to r88 with the new release and check if sound still works ?

Regarding nested kvm, Google only enabled it on devices with 4.19 kernel and above. I have no idea what is the reason behind it (kernel requirements or something else) but you can try to add a line with "kvm_host" text to /etc/ui_use_flags.txt, reboot and see if it is enabled in crostini. However, if it works you might be better with the official firmware and just dev mode.

Thanks for the donation offer, I really appreciate the gesture :) I would prefer that my work helps charity/hospitals, don't hesitate to make a small donation to one you support if you feel like it.

**Reply by hopgoodpaul** [reply]
After trying your latest build with 4.4 kernel, I can confirm all is working except for the camera (as expected). Bluetooth appears to work as well.

It's frustrating that Google don't update the kernel very often at all. I had the c302 for a long time waiting for linux support and now I'm in the same situation waiting for nested kvm support. I will give r88 a try too.

When I check /etc/ui_use_flags.txt on your build, I can see "kvm_host" is already in there but checking 
"cat /sys/module/kvm_intel/parameters/nested" it returns N. The same command in Crostini shows Y but /dev/kvm isn't present whereas /dev/kvm is present on the chronos shell. Am I misunderstanding how nested kvm should appear? Installing Windows 10 is painfully slow and doesn't actually complete (or perhaps my patience running out doesn't allow it to complete).

Using device native kernel was what originally got me thinking about using Crunch. Would that allow me to enable nested kvm?

**Reply by sebanc** [reply]
You actually know more than me about this, try adding "kvm_intel.nested=1" on the kernel command line.

Crunch would not help in this case, it is the exact same ChromeOS as the original device (kernel+rootfs) booting from efi.

**Reply by sebanc** [reply]
As a side note, to launch a windows vm in brunch I would probably add "enforce_hyperthreading=1"on the kernel command line and use "brioche" (https://github.com/sebanc/brioche).
